astropy/coordinates/baseframe.py

```diff
@@ -337,36 +337,17 @@ def __init__(
                 f"keywords: {list(kwargs)}"
             )
 
-        # We do ``is None`` because self._data might evaluate to false for
-        # empty arrays or data == 0
-        if self._data is None:
-            # No data: we still need to check that any non-scalar attributes
-            # have consistent shapes. Collect them for all attributes with
-            # size > 1 (which should be array-like and thus have a shape).
-            shapes = {
-                fnm: value.shape
-                for fnm, value in values.items()
-                if getattr(value, "shape", ())
-            }
-            if shapes:
-                if len(shapes) > 1:
-                    try:
-                        self._no_data_shape = check_broadcast(*shapes.values())
-                    except ValueError as err:
-                        raise ValueError(
-                            f"non-scalar attributes with inconsistent shapes: {shapes}"
-                        ) from err
-
-                    # Above, we checked that it is possible to broadcast all
-                    # shapes.  By getting and thus validating the attributes,
-                    # we verify that the attributes can in fact be broadcast.
-                    for fnm in shapes:
-                        getattr(self, fnm)
-                else:
-                    self._no_data_shape = shapes.popitem()[1]
-
-            else:
-                self._no_data_shape = ()
+        shapes = [
+            value.shape for value in values.values() if getattr(value, "shape", ())
+        ]
+        if self._data is not None and getattr(self._data, "shape", ()):
+            shapes.append(self._data.shape)
+        try:
+            self._shape = check_broadcast(*shapes)
+        except ValueError as err:
+            raise ValueError(
+                f"non-scalar attributes with inconsistent shapes: {shapes}"
```

Comment:
This error message is a bit confusing because it is also raised when there is a mismatch between the shape of `data` and the shape of frame attribute(s).

Reply:
How about:

```suggestion
                f"non-scalar data and/or attributes with inconsistent shapes: {shapes}"
```

Reply:
Fixed.